src/PowerShellEditorServices/Language/RunspaceSynchronizer.cs

```diff
@@ -0,0 +1,272 @@
+            targetEngineIntrinsics = ReflectionUtils.GetEngineIntrinsics(runspaceTarget);
+            IsReadyForEvents = true;
+
+            sourceEngineIntrinsics.Events.SubscribeEvent(
+                source: null,
+                eventName: null,
+                sourceIdentifier: PSEngineEvent.OnIdle.ToString(),
+                data: null,
+                handlerDelegate: HandleRunspaceStateChange,
+                supportEvent: true,
+                forwardEvent: false);
+
+            Activate();
+            // Trigger events
+            HandleRunspaceStateChange(sender: null, args: null);
+        }
+
+        /// <summary>
+        /// Does a thing
+        /// </summary>
+        public static void Activate()
+        {
+            SourceActionEnabled = true;
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private static void HandleRunspaceStateChange(object sender, PSEventArgs args)
+        {
+            if (!SourceActionEnabled)
+            {
+                return;
+            }
+
+            SourceActionEnabled = false;
+
+            var newOrChangedModules = new List<PSModuleInfo>();
+            List<PSModuleInfo> modules = ReflectionUtils.GetModules(sourceRunspace);
+            foreach (PSModuleInfo module in modules)
+            {
+                if (moduleCache.Add(module))
+                {
+                    newOrChangedModules.Add(module);
+                }
+            }
+
+
+            var newOrChangedVars = new List<PSVariable>();
+
+            var variables = sourceEngineIntrinsics.GetVariables();
+            foreach (var variable in variables)
+            {
+                // If the variable is a magic variable or it's type has not changed, then skip it.
+                if(POWERSHELL_MAGIC_VARIABLES.Contains(variable.Name) ||
+                    (variableCache.TryGetValue(variable.Name, out Type value) && value == variable.Value?.GetType()))
+                {
+                    continue;
+                }
+
+                // Add the variable to the cache and mark it as a newOrChanged variable.
+                variableCache[variable.Name] = variable.Value?.GetType();
+                newOrChangedVars.Add(variable);
+            }
+
+            if (newOrChangedModules.Count > 0)
+            {
+                // Import the modules in the targetRunspace with -Force
+                using (PowerShell pwsh = PowerShell.Create())
+                {
+                    pwsh.Runspace = targetRunspace;
+
+                    foreach (PSModuleInfo moduleInfo in newOrChangedModules)
+                    {
+                        if(moduleInfo.Path != null)
+                        {
+                            pwsh.AddCommand("Import-Module")
+                                .AddParameter("Name", moduleInfo.Path)
+                                .AddParameter("Force")
+                                .AddStatement();
```

Comment:
Wouldn't this make the psm1 run twice?

Also in Windows PowerShell, static class methods have runspace affinity to the last runspace that ran the type definer.  This second import will change the runspace that static class methods run in, even when invoked in the origin runspace.

Reply:
This affinity issue has been fixed in PowerShell Core. See https://github.com/PowerShell/PowerShell/pull/4209

Reply:
Yeah I did qualify it, I should have been more clear though.  That said, afaik there aren't plans to drop support for it anytime soon right?

Reply:
No, there is no plan to port the fix back to Windows PowerShell.

Reply:
What I'm getting at is PSES still has to support Windows PowerShell.  So it seems to me that the choices are:

1. Break static methods in Windows PowerShell
1. Maintain both this implementation and the current system
1. Stop supporting Windows PowerShell

Or am I missing something?

Reply:
Or, 4. Only apply this optimization to PowerShell Core. On Windows PowerShell, keep calling into the integrated console for intellisense operations

Reply:
Just to clarify, that's what I meant by item 2.

Reply:
I'm currently now doing 2. If you're using PowerShell Core, you will get this optimization.